app/src/main/java/com/example/farmeraid/data/MarketRepository.kt

```diff
@@ -2,69 +2,95 @@ package com.example.farmeraid.data
+    }
+
+    private suspend fun readMarketDataWithQuotas(): ResponseModel.FAResponseWithData<List<MarketModel.MarketWithQuota>> {
+        val marketIds = farmRepository.getMarketIds()
+        if (marketIds.error != null) {
+            return ResponseModel.FAResponseWithData.Error(marketIds.error)
+        }
 
-    fun getMarkets(): Flow<List<MarketModel.Market>> {
+        val marketModel : MutableList<DocumentSnapshot> = mutableListOf()
+        marketIds.data?.forEach {
+            try {
+                marketModel.add(db.collection("market").document(it).get().await())
+            } catch (e : Exception) {
+                return ResponseModel.FAResponseWithData.Error(e.message ?: "Unknown error while fetching market")
+            }
+        }
+
+        val marketModelList = mutableListOf<MarketModel.MarketWithQuota>()
+
+        for (market in marketModel) {
+            quotasRepository.getQuota(market.id)?.let {
+                MarketModel.MarketWithQuota(
+                    id = market.id,
+                    name = market.get("name") as String,
+                    quota = it
+                )
+            }?.let { marketModelList.add(it) }
+        }
+
+        return ResponseModel.FAResponseWithData.Success(marketModelList as List<MarketModel.MarketWithQuota>)
+    }
+
+    suspend fun getMarkets(): Flow<ResponseModel.FAResponseWithData<List<MarketModel.Market>>> {
         return flow {
-            emit(markets)
+            emit(readMarketData())
         }.flowOn(Dispatchers.IO)
     }
 
-    fun getMarketsWithQuota() : Flow<List<MarketModel.MarketWithQuota>> {
+    suspend fun getMarketsWithQuota() : Flow<ResponseModel.FAResponseWithData<List<MarketModel.MarketWithQuota>>> {
         return flow {
             emit(
-                markets.mapNotNull { market ->
-                    quotasRepository.getQuota(market.id)
-                        ?.let { quota ->
-                            MarketModel.MarketWithQuota(
-                                id = market.id,
-                                name = market.name,
-                                quota = quota,
-                            )
-                        }
-                }
+                readMarketDataWithQuotas()
             )
         }
     }
 
-    fun getMarket(id: String) : MarketModel.Market? {
-        return markets.firstOrNull { it.id == id }
-    }
-
     suspend fun getMarketWithQuota(id : String) : MarketModel.MarketWithQuota? {
-        return markets.firstOrNull { it.id == id }
-            ?.let { market ->
-                quotasRepository.getQuota(market.id)
-                    ?.let { quota ->
-                        MarketModel.MarketWithQuota(
-                            id = market.id,
-                            name = market.name,
-                            quota = quota,
-                        )
-                    }
-            }
+        return readMarketDataWithQuotas().data?.firstOrNull { it.id == id }
```

Comment:
It's possible for readMarketDataWithQuotas to return an error. Therefore, return an `FAResponseWithData<MarketModel.MarketWithQuota?>` in this function. And for the return put:
```
return readMarketDataWithQuotas().let { marketWithQuotas ->
    marketWithQuotas.data?.let {
        ResponseModel.FAResponseWithData.Success(it.firstOrNull { it.id = id })
    } ?:  ResponseModel.FAResponseWithData.Error(marketWithQuotas.error ?: "Unkown error reading quotas")
}
```